src/NuGetGallery/Controllers/UsersController.cs

```
@@ -1029,6 +1034,308 @@ public virtual ActionResult PasswordChanged()
             return View();
         }
 
+        [HttpGet]
+        [UIAuthorize]
+        public virtual ActionResult TrustedPublishing()
+        {
+            User currentUser = GetCurrentUser();
+            if (!_featureFlagService.IsTrustedPublishingEnabled(currentUser))
+            {
+                Response.StatusCode = (int)HttpStatusCode.BadRequest;
+                return Json(Strings.DefaultUserSafeExceptionMessage);
+            }
+
+            var userPolicies = _federatedCredentialRepository.GetPoliciesCreatedByUser(currentUser.Key);
+
+            // Show newest policies on the top.
+            var policies = userPolicies
+                .OrderByDescending(p => p.Created)
+                .Select(p => CreatePublisherViewModel(currentUser, p))
+                .Where(p => p != null)
+                .ToArray();
+
+            var owners = new List<User>() { currentUser };
+            owners.AddRange(currentUser.Organizations.Select(o => o.Organization));
+
+            var model = new TrustedPublisherPolicyListViewModel
+            {
+                Username = currentUser.Username,
+                Policies = policies,
+                PackageOwners = owners.Select(o => o.Username).ToArray(),
+            };
+
+            return View("TrustedPublishing", model);
+        }
+
+        private TrustedPublisherPolicyViewModel CreatePublisherViewModel(User currentUser, FederatedCredentialPolicy policy)
+        {
+            if (policy.Type != FederatedCredentialType.GitHubActions)
+            {
+                return null;
+            }
+
+            if (GitHubPolicyDetailsViewModel.FromDatabaseJson(policy.Criteria)
+                is not GitHubPolicyDetailsViewModel policyDetails)
```

I almost think we should 5XX (throw an exception either manually or by JSON deserialization error from `FromDatabaseJson`, in this case instead of silently failing. This is definitely a bug if the JSON on a `GitHubActions` type policy can't be deserialized. What do you think?

---

Ah, got it—so throwing is the preferred way to flag internal issues like this. Good to know, thanks!

---

Yes, we get HTTP 5XX logs in App Insights, users will see an error and notify us. This is not ideal of course, but better than silently hiding data IMO.